### ISSUE TITLE
Rewriting Lebab as a Babel transformer?

### ISSUE THREAD

**Issue by mohebifar** [mohebifar/lebab#138]
Currently lebab is doing so much from parsing to generating code. During the process, somethings are missing (Like nice warnings while parsing the code #136 or JSX Support #130) or even some bugs show up. I believe almost all of these issues have been considered and solved in babel.

Lebab's aim is basically transforming an old code to a modern code. That means the only thing that matters in lebab is transforming. So, making a babel transformer instead of re-implementing all that stuff can be an option for the future releases.

**Comment by nene** [comment]
That's an intriguing idea, but I to my understanding Babel does not fully preserve the original whitespace. For example, when transforming the following code, that has nothing to transform:

```js
/**
 * My func
 */
function  foo   (y) {
    // Hello world
    if (  true  ) {
            // poor indentation
    }
}
```

Babel produces:

```js
/**
 * My func
 */
function foo(y) {
  // Hello world
  if (true) {
    // poor indentation
  }
}
```

It does preserve all the comments, but it pretty much reformats all the whitespace. This behavior is fine for Babel, as one only cares that the output is readable.

In Lebab's case you'll want the output to be 100% the same as input when no transforms were done - so Lebab would not mess up whatever strange coding convention one has, and you could look at the diff of code before/after and only see the actual transformations without any whitespace noise.

So it seems to me that this is not a feasible approach until Babel adopts a code generator more similar to Recast.

**Comment by nene** [comment]
On further investigation I discovered that it might be feasible to use the parser of Babel - Babylon. The AST it generates is mostly compatible with ESTree format used by Esprima and others, but there are some [minor differences.](https://github.com/babel/babylon/issues/40)

**Comment by benjamn** [comment]
Recast is parser-agnostic; you can call `recast.parse(source, { parser: whatever })`, as long as `whatever` is an object with a `.parse` method that produces AST nodes with `.loc` information.

**Comment by nene** [comment]
@benjamn Yep, Lebab is already using that run Espree parser instead. The problem with Babylon is that it doesn't produce AST that's fully compatible with ESTree spec.

**Comment by hzoo** [comment]
@benjamn So we could switch out `babel-generator` for recast.print to keep original formatting (assuming you wanted something like jscodeshift/eslint autofix/run babel like `babel src --out-dir src`?

**Comment by BrendanAnnable** [comment]
@hzoo are we not already using `recast.print`? https://github.com/mohebifar/lebab/blob/master/src/transformer.js#L59

**Comment by hzoo** [comment]
A babel transform/babel uses babylon + babel-traverse + babel-generator so you could modify the ast with a babel transform and output to something that can consume a babel AST?

**Comment by benjamn** [comment]
Well, Babel is optimized for code generation performance, and preserving formatting takes a bit more effort, but `recast.print` will work with any AST that `recast.parse` parsed (as long as the nodes have `.loc.{start,end}` information).

**Comment by hzoo** [comment]
Ok I'm just wondering what we can do to make recast + babel interop better if that's possible. I think it would be nice to be able to write a babel transform like `babel src -d src --presets lebab` and it uses recast instead of babel-generator to preserve source (codemod, autofix, etc).

**Comment by benjamn** [comment]
If this works (and I don't know why it wouldn't), then it shouldn't be hard:
```js
var recast = require("recast");
var ast = recast.parse(source, { parser: require("babylon") });
var result = require("babel-core").transformFromAst(ast, {
  presets: ["lebab"],
  ast: true
});
console.log(recast.print(result.ast).code);
```

**Comment by hzoo** [comment]
@benjamn would recast have to parse it like `recast.parse(source, { parser: require("babylon") });` or can you still use babylon on its own? Just wanted to know what changes would need to be made to babel to support this for a regular babel setup (sounds like we would need an option to use a different parser/generator in babelrc)?

**Comment by benjamn** [comment]
Letting `recast.parse` do the parsing is not essential, in principle, but it does have several advantages:
* The `recast.parse` code can pass [the options it needs](https://github.com/benjamn/recast/blob/5c6bf309e7d5bb7618e04b3fd4b07010537b1c0f/lib/parser.js#L26-L35) to the parser, reducing manual configuration.
* The `.loc.{start,end}` information in the parsed AST is guaranteed to correspond to the original source, since no other code has had the chance to modify the AST. In some cases `recast.parse` even [fixes bugs](https://github.com/benjamn/recast/blob/5c6bf309e7d5bb7618e04b3fd4b07010537b1c0f/lib/parser.js#L107) related to `.loc` information for the various Esprima-like parsers that it allows.
* The real magic of `recast.parse` is that it makes a [copy](https://github.com/benjamn/recast/blob/5c6bf309e7d5bb7618e04b3fd4b07010537b1c0f/lib/parser.js#L87) of the AST, and links every node in the copy to its original node via `.original` properties. This shadow-AST is what enables `recast.print` to perform tree diffing and reuse original source code when possible.

In other words, you could parse the AST yourself, and then call some other `recast.*` API to make the copy and create the `.original` linkage, but `recast.parse` is a simpler interface for that, I think.

**Comment by hzoo** [comment]
Ok I can test it now my modifying babel's transformation step to take in a different parser/generator. I think the last thing is to accept babel AST nodes in ast-types? Since `Error: unknown: {type: ObjectProperty} does not match type Printable`

**Comment by hzoo** [comment]
@benjamn yay something like this seems to work https://github.com/babel/babel/pull/3561

**Comment by hzoo** [comment]
As a update for this to happen:

- [ ] https://github.com/babel/babel/pull/3561 in babel-core needs to land in some form (probably a option to pass a string (similar to a plugin) that resolves to a module with a parse method and a generator method) - I'm testing in my local astexplorer
- [ ] https://github.com/benjamn/ast-types/pull/162 update ast-types with babel 6 types
- [ ] https://github.com/benjamn/recast/pull/299 update recast printer with the same types

**Comment by hzoo** [comment]
https://github.com/babel/babel/pull/3561 has landed! So writing a plugin that doesn't include the different AST nodes should work now https://github.com/babel/babylon#output

**Comment by hzoo** [comment]
Merged the 2 prs ^ so this is actually a possibility now! example here: https://github.com/fkling/astexplorer/pull/161

**Comment by hzoo** [comment]
Let me know if you want to discuss doing this at some point again! I know there's a lot on everyone's plate (and it will take a good amount of work) but just checking the option open (now that this can be done). Can also chat on our slack: http://slack.babeljs.io/

Could be a good opportunity for some simple beginner-friendly contribution issues as well if everything is planned out. cc @nene @mohebifar

**Comment by nene** [comment]
I actually started out with switching the parser in Lebab over to Babylon. Wrote an adapter to convert Babylon node types to Esprima node types, this got most of the tests passing, but that's really a temporary solution. So got a bit stuck behind the task of converting the code really over to use Babylon AST.

Thanks for the resources!

**Comment by hzoo** [comment]
Yeah definetely reach out if you have any questions/concerns: I know there's a lot to know and yes its certainly overwhelming. 

> Wrote an adapter to convert Babylon node types to Esprima node types, this got most of the tests passing, but that's really a temporary solution

I actually had to do this exact thing for https://github.com/babel/babel-eslint if you didn't know heh https://github.com/babel/babel-eslint/tree/master/babylon-to-espree

And also FYI we are also starting out an effort to make a plugin to output back to ESTree/Babel 5 output in the parser. https://github.com/babel/babylon/pull/277 (@danez) which is basically the same logic but built-into the parser. The way we did it in babel-eslint is slow and really bad since we parse with babylon first and then change nodes around to fit the other format.

**Comment by nene** [comment]
I wonder how well does something like this babylon-to-espree work? Like what are the downsides of not using Babylon AST directly? For Lebab the speed isn't really a crucial aspect. Perhaps I'm better off with an adapter layer like this.

I guess my main concern is coupling Lebab too tightly with the non-standard AST tree of Babylon, which is likely to fluctuate more than ESTree spec. Though I have to say I like some of the AST nodes of Babylon much more than their ESTree equivalents (especially the separate Literal nodes).

**Comment by nene** [comment]
Rewriting Lebab transforms as Babel transforms is another direction I'd really like to explore. But that will be a larger undertaking. I've accumulated a fair chunk of code in all these transforms. But I've also come across some of the limitations of the current system - especially with a need to track variables. Some sort of evolutionary approach would be nice to gradually reimplement the transforms as Babel-transforms, while keeping the others working alongside.

**Comment by hzoo** [comment]
Yeah the only downsides are speed plus hard to maintain + not enough tests (just a time/resource issue).

> I guess my main concern is coupling Lebab too tightly with the non-standard AST tree of Babylon, which is likely to fluctuate more than ESTree spec. 

We aren't planning on making anymore changes

> Though I have to say I like some of the AST nodes of Babylon much more than their ESTree equivalents (especially the separate Literal nodes).

Yeah the changes I think are definetely nice especially for transforming which makes sense for Babel and for the AST, but in the end it doesn't interop well unless everyone agrees on doing ESTree 2.0

**Comment by hzoo** [comment]
Is this still something we want to do for Lebab? I was wondering if we could ask this ask as part of https://developers.google.com/open-source/gsoc/?

**Comment by nene** [comment]
I do still have very much interest in doing this. Though I have drawn myself back from various open-source developments lately to just spend more time in the real world.

Anyway, I have nothing against proposing this task for Google Summer of Code. It might even be better this way as I've found I'm better at cleaning up the code others have written than writing new stuff by myself from scratch.

**Comment by hzoo** [comment]
> I do still have very much interest in doing this. 

Cool 😄

> Though I have drawn myself back from various open-source developments lately to just spend more time in the real world.

Totally understandable and a good thing!

Awesome, yeah I add it to the list, and hopefully we get some volunteers on this effort then.

We'll just want to plan out how it should happen (I think we would all want it to be incremental so it's easy to review)

**Comment by rajatvijay** [comment]
Hi guys, I would like to work on this project under GSOC. Lately I have been learning AST and how to write plugins for babel. 
Any help like even pointing out to resources to get started will be of great cause. Thanks in advance!

**Comment by nene** [comment]
Hard to point out any general resources. If you have a specific question in mind, feel free to ask.

**Comment by rajatvijay** [comment]
@nene thanks!
So I have tried Babel and lebab both, on the product level and went through the source code of lebab.
I think the starting point should be deciding the specs of all the babel plugins that will be developed.

If you have any other starting point in mind please let me know.

**Comment by nene** [comment]
The approach I originally tried regarding this whole thing was to replace just the parser used by Lebab with Babel. It didn't really turn out quite as simple though, but I guess it would still be a feasible approach. Another side of the coin is that AST manipulation inside current Lebab code has several shortcomings. The main shortcomings are in dealing with variable scoping. This is where I think switching over to Babel API-s should improve Lebab the most.

Regarding a spec, I think the functionality of Lebab is pretty well specified by the tests. So when re-implementing Lebab functionality as a Babel plugin, one mainly needs to ensure that all the existing tests pass.

**Comment by rajatvijay** [comment]
So the main project is to improve the lebab or rewrite it as a babel plugin?

**Comment by nene** [comment]
I think the transforms that Lebab does could be rewritten as a Babel plugins. But it would still be nice to package them inside the Lebab program, so that nothing would change for somebody already using Lebab. This way we could also replace the functionality gradually. There's quite a bit of code in Lebab. Better to tackle it one transform at a time. Of course, the Babel plugins could additionally be used separately from Lebab, if one wishes so, but I think for most people it would be simpler to use as a single program.

**Comment by rajatvijay** [comment]
Okay so I think I should go forward with rewriting the lebab transforms as a babel plugin **one by one**, starting with the **let** transform. Also, will make sure that the plugins can be packaged in a lebab program so the legacy functionality doesn't break.

Please feel free to point out if there is anything that I missed. Thanks in advance!

**Comment by nene** [comment]
Sounds great.

**Comment by hzoo** [comment]
@rajatvijay If you want to do this as part of gsoc, you need to check out https://summerofcode.withgoogle.com/organizations/5842528113786880/ and the gsoc website to apply with your proposal and other information.

**Comment by rajatvijay** [comment]
@hzoo but the applications start on March 20 right?

**Comment by rajatvijay** [comment]
@nene, @hzoo  Hi!
I started with writing the babel plugin to transform require calls to import declarations. Have made some progress in accordance with the tests written in lebab, would say 50% through it. 
But I guess getting a code review of what I have written so far will have me straighten up my line of thought and also get a sense of a quality of my style coding, as it suits the lebab/babel community or not. 
Also, I don't think generating a pull request will be a good idea since the plugin is not complete.

**Comment by nene** [comment]
I'm sure I could review your code and provide some feedback.

A pull request is fine, even when it's completely broken, there's no requirement to actually merge it. Or present the code in some other form that you find more suitable.

**Comment by rajatvijay** [comment]
@nene Here is the link to the file that contains the plugin code. 

[https://github.com/rajatvijay/babel-plugin-transform-require/blob/master/src/index.js](url)

I completely understand that it can be refactored and improved, but since it's not a complete implementation, I have majorly focussed in passing the tests.

Also made it a gist so that you can provide your comments there and then. The gist url is:
[https://gist.github.com/rajatvijay/852003cb37afea370ad459c6f3de84c1#file-index-js-L1](url)

**Comment by rajatvijay** [comment]
@nene Hi
Did you get a chance to see the code?

**Comment by nene** [comment]
Hi there. I was busy slacking off during weekend :)

Now finally found some time to review your code. Wrote quite a lot of comments. Unfortunately Github gists don't allow writing inline comments for particular lines of code. Tried to make my best without that.

See: https://gist.github.com/nene/d9d9ec9802fd6e38cb4e212b8ca90b84

**Comment by nene** [comment]
Also, one think that we should consider rather sooner than later is the fact that ES6 imports are effectively `const`, so the following code should not be transformed:

```js
var foo = require("foo");
...
foo = somethingElse();
```

Dealing with variable scoping is currently one of the main weaknesses of Lebab. Babel should provide us with much better scope information.

**Comment by rajatvijay** [comment]
Hi @nene 
Very very thanks for the review. It changed my way of looking at the problem in hand.

Sorry for replying so late. Got busy in some college exams.

Improving upon the points mentioned. And really like the "patterns matching in trees" blog post. It can help reduce a lot of code. 

Will start off by extracting that feature as a separate library from lebab so can be used in the current problem and such problems coming forth.

**Comment by rajatvijay** [comment]
@nene 

I have one doubt.

> Dealing with variable scoping is currently one of the main weaknesses of Lebab. Babel should provide us with much better scope information.

Why need a better scope information if babel API handles it well itself?

**Comment by nene** [comment]
What I meant was that scope information provided by Babel seems to be better than the one we currently in Lebab get through a separate helper library.

**Comment by rajatvijay** [comment]
Hi @nene 

Apologies for disappearing had some personal crises.

Wanted to confirm if this task is still up for grabs or can I resume working on it, irrespective of any summer of code challenge.

 Thanks

**Comment by nene** [comment]
Hi @rajatvijay, you are very much free to continue working on this.

Barely anything has changed in the mean time.

**Comment by hzoo** [comment]
Yep, can totally be done outside of any program, was just a suggestion. Maybe can look at https://github.com/square/babel-codemod as well by @eventualbuddha

**Comment by hzoo** [comment]
This would still be amazing to do, for RGSoC we have a codemod for a TC39 proposal WIP which also could be a good candidate for this https://github.com/babel/babel/pull/6048

**Comment by abiduzz420** [comment]
Hello @nene @mohebifar I find lebab to be a really handy tool for doing large code base migrations from 5to6. I would like to talk to you on the subject of using lebab as a babel preset.
I have few questions which could help us understand about the idea better: 

I would like to know if this idea be implemented ?  Can we make the use of tool better, if we internally integrate with babel without the user having to configure anything extra. Since this would involve re-writing of majority of transformations, I wish to hear your thoughts before moving further.

Thanks in advance :)

**Comment by nene** [comment]
Hi @abiduzz420 I don't quite understand what would be the benefit of implementing Lebab as a Babel preset. As I see it, to use a babel preset you would actually need to configure more than you currently need to do with Lebab (really, no configuration at all). And they way you use Babel is pretty different from the way you would use Lebab. You would use Lebab to perform the migration once, after which you won't have much use for it any more. In contrast to Babel that you'd keep continuously running. Also with Lebab you usually want the transforms to modify the existing files, versus Babel, where you'd rather output the generated sources to another file.

Perhaps I misunderstood you. I have nothing against switching to Babylon parser and making use of some other existing Babel-related tools.

**Comment by hzoo** [comment]
I tried to explain this in https://github.com/babel/babel/issues/7296#issuecomment-361648387 and here already but I only thought we wanted to change Lebab internally to use babel, the cli would be the same, no configuration needed etc. Technically this could simply just mean creating each transform as a plugin (with a preset) and to turn on or turn off each plugin (or plugin option) by mapping those to the cli options.

And the purpose of this is to support newer syntax in the parser as well as take advantage of tooling used in babel. The user doesn't need to know this change happens so it's basically a refactor. Lebab could just be a wrapper around a preset and run it (think of it as another babel-cli)

**Comment by abiduzz420** [comment]
Yes, this suddenly makes a lot of sense to me now (using it as another babel-cli). I did not make myself clear when I said plugin based transforms and keeping the cli intact.
I understand the problem statement better now.

**Comment by eventualbuddha** [comment]
I would be very interested in seeing whether [babel-codemod](https://github.com/square/babel-codemod) would work as a base for “lebab as a preset”. It should be fairly easy to do, and I can put together an example runner with an existing babel preset if that would help.

**Comment by rajatvijay** [comment]
Hey @nene @hzoo 

I have put together a babel plugin to transform require calls to import calls. Here is the repo for it.
https://github.com/rajatvijay/babel-plugin-transform-require.

The plan is to compose lebab from such babel plugins and wrap it inside the lebab cli utility for backward compatibility.

Let me know in what ways [this](https://github.com/rajatvijay/babel-plugin-transform-require) plugin can be improved and should I move ahead in this direction to make others plugins too.

**Comment by nene** [comment]
@rajatvijay This looks like a great start!

A few things that immediately pop to my mind:

- We should copy the existing Lebab tests over and make them work with this plugin. Replacing the manually runnable tests in try/ dir with automated ones.
- The `matchesAst` and friends should really be moved to a package of its own. I've been planning to do that, but haven't so far had a good motivation for doing so. I guess now it's finally time for that.
- The repo also contains the compiled sources in lib/. That's usually not a good idea.

**Comment by nene** [comment]
@rajatvijay I extracted the `matchesAst()` helper out of Lebab and published in npm under name [f-matches](https://github.com/nene/f-matches).

While doing so I changed the API slightly:

- The module exports three functions: `matches`, `matchesLength`, `extract`.
- So mainly I've dropped the AST-part from function name, as there really isn't nothing specific to AST nodes there.
- `isAstMatch()` has been eliminated by making `matches()` a curried function. Which means that the argument order is swapped now.
- The two other functions are also curried.

In short, you should be able to just replace `matchesAst` with `matches` imported from `f-matches` package.

**Comment by rajatvijay** [comment]
@nene 

> The repo also contains the compiled sources in lib/. That's usually not a good idea.

This has been fixed.

> I extracted the matchesAst() helper out of Lebab and published in npm under name f-matches.

This is going to be my next step.
 
1. Integrate this library into the plugin 
2. and then will migrate the tests too

**Comment by rajatvijay** [comment]
@nene the plugin has been published to npm with all the tests passing and using f-matches.

Moving on to the new transform.

**Comment by eventualbuddha** [comment]
Might I suggest following babel’s Lead and using a monorepo for the lebab-as-babel-plugins packages?

**Comment by niieani** [comment]
I recommend using `yarn` workspaces instead of things like `lerna`. It does almost everything necessary and works much better overall in my experience.

**Comment by eventualbuddha** [comment]
This is an explanation of how we use a monorepo with yarn and Ember. Not that much is Ember-specific.

**Comment by rajatvijay** [comment]
@eventualbuddha @niieani 
I have pushed the monorepo here [lebab-as-babel-plugins](https://github.com/rajatvijay/lebab-as-babel-plugins). 

But due to yarn workspaces `jest` is not able to find `babel-jest`. 
Here is an active issue https://github.com/facebook/react-native/issues/17469

**Comment by rajatvijay** [comment]
Sorry got the solution for this

> But due to yarn workspaces jest is not able to find babel-jest.
Here is an active issue facebook/react-native#17469

`.babelrc` was missing in the root folder

thanks

**Comment by rajatvijay** [comment]
Thanks @niieani  and @eventualbuddha. I have successfully made a monorepo for this, [here](https://github.com/rajatvijay/lebab-as-babel-plugins)

@nene was working on transform-export plugin and was looking at the tests and got confused: 

```javascript
it('should ignore function export when function name does not match with exported name', () => {
      expectNoChange('exports.foo = function bar() {};');
});
```

Shouldn't this be transformed to `export {bar as foo}`?
 [Here](https://github.com/lebab/lebab/blob/master/test/transform/commonjs/exportCommonjsTest.js#L54) is the link to these tests.

**Comment by eventualbuddha** [comment]
> Shouldn't this be transformed to `export {bar as foo}`?

Maybe. It depends on what else is in the file. If that were the only thing, then yes, and [that's what esnext does](https://esnext.github.io/esnext/#exports.foo%20%3D%20function%20bar()%20%7B%7D).

As long as you aren't changing the behavior of the program unexpectedly, you should make what upgrades you can. In this case, as long as there is no existing `bar` in scope or any references to a global `bar`, you can make `function bar() {}` be a `FunctionDeclaration` instead of a `FunctionExpression` and then use the export declaration you indicated.

I recommend checking out the [esnext tests](https://github.com/esnext/esnext/tree/master/test/form/modules.commonjs) and possibly [implementation](https://github.com/esnext/esnext/blob/master/src/plugins/modules.commonjs.js) to help you.
